client/internal/clientcommon.go

```diff
@@ -25,11 +27,105 @@ var (
 	errReportsPackageStatusesNotSet = errors.New("ReportsPackageStatuses capability is not set")
 )
 
+// CallbacksWrapper wraps Callbacks such that it is possible to query if any callback
+// function is in progress (called, but not yet returned). This is necessary for
+// safe handling of certain ClientCommon methods when they are called from the callbacks.
+// See for example Stop() implementation.
+type CallbacksWrapper struct {
+	wrapped types.Callbacks
+	// Greater than zero if currently processing a callback.
+	inCallback *int64
+}
+
+func (cc *CallbacksWrapper) OnConnect() {
+	cc.EnterCallback()
+	defer cc.LeaveCallback()
+	cc.wrapped.OnConnect()
+}
+
+func (cc *CallbacksWrapper) OnConnectFailed(err error) {
+	cc.EnterCallback()
+	defer cc.LeaveCallback()
+	cc.wrapped.OnConnectFailed(err)
+}
+
+func (cc *CallbacksWrapper) OnError(err *protobufs.ServerErrorResponse) {
+	cc.EnterCallback()
+	defer cc.LeaveCallback()
+	cc.wrapped.OnError(err)
+}
+
+func (cc *CallbacksWrapper) OnMessage(ctx context.Context, msg *types.MessageData) {
+	cc.EnterCallback()
+	defer cc.LeaveCallback()
+	cc.wrapped.OnMessage(ctx, msg)
+}
+
+func (cc *CallbacksWrapper) OnOpampConnectionSettings(
+	ctx context.Context, settings *protobufs.OpAMPConnectionSettings,
+) error {
+	cc.EnterCallback()
+	defer cc.LeaveCallback()
+	return cc.wrapped.OnOpampConnectionSettings(ctx, settings)
+}
+
+func (cc *CallbacksWrapper) OnOpampConnectionSettingsAccepted(settings *protobufs.OpAMPConnectionSettings) {
+	cc.EnterCallback()
+	defer cc.LeaveCallback()
+	cc.wrapped.OnOpampConnectionSettingsAccepted(settings)
+}
+
+func (cc *CallbacksWrapper) SaveRemoteConfigStatus(ctx context.Context, status *protobufs.RemoteConfigStatus) {
+	cc.EnterCallback()
+	defer cc.LeaveCallback()
+	cc.wrapped.SaveRemoteConfigStatus(ctx, status)
+}
+
+func (cc *CallbacksWrapper) GetEffectiveConfig(ctx context.Context) (*protobufs.EffectiveConfig, error) {
+	cc.EnterCallback()
+	defer cc.LeaveCallback()
+	return cc.wrapped.GetEffectiveConfig(ctx)
+}
+
+func (cc *CallbacksWrapper) OnCommand(command *protobufs.ServerToAgentCommand) error {
+	cc.EnterCallback()
+	defer cc.LeaveCallback()
+	return cc.wrapped.OnCommand(command)
+}
+
+var _ types.Callbacks = (*CallbacksWrapper)(nil)
+
+func NewCallbacksWrapper(wrapped types.Callbacks) *CallbacksWrapper {
+	zero := int64(0)
+
+	if wrapped == nil {
+		// Make sure it is always safe to call Callbacks.
+		wrapped = types.CallbacksStruct{}
+	}
+
+	return &CallbacksWrapper{
+		wrapped:    wrapped,
+		inCallback: &zero,
+	}
+}
+
+func (cc *CallbacksWrapper) EnterCallback() {
+	atomic.AddInt64(cc.inCallback, 1)
```

Comment:
At the moment we don't really have concurrent execution of callbacks, so we could just use plain non-atomic increment/decrement, but it is just trivial to use atomics here so that in case in the future we start calling callbacks concurrently we don't end up forgetting to fix this code. There is virtually no performance impact since callback are called very rarely.